crates/release_plz_core/src/published_packages.rs

```diff
@@ -0,0 +1,274 @@
+                    let mut downloader =
+                        download::PackageDownloader::new(packages_names, directory);
+                    if let Some(registry) = registry {
+                        downloader = downloader.with_registry(registry.to_string());
+                    }
+                    registry_packages.extend(
+                        downloader
+                            .download()
+                            .context("failed to download packages")?,
+                    );
+                }
+
+                // After downloading the package, we initialize a git repo in the package.
+                // This is because if cargo doesn't find a git repo in the package, it doesn't
+                // show hidden files in `cargo package --list` output.
+                let registry_packages = initialize_registry_package(registry_packages)
+                    .context("failed to initialize repository package")?;
+                self.extend(registry_packages);
+            }
+        }
+        Ok(())
+    }
+
+    /// Retrieves the latest [`PublishedPackage`]s for the given `packages`
+    /// from git tags in the `repo`.
+    ///
+    /// This function upon completion may leave the `repo` in a different state
+    /// than when it was called; it is the caller's responsibility to restore it to the
+    /// original state if needed (for example, with [`Repo::checkout_head`]).
+    fn get_latest_tagged_packages<'p>(
+        &mut self,
+        project: &Project,
+        repo: &Repo,
+        packages: impl Iterator<Item = &'p Package> + 'p,
+    ) -> anyhow::Result<()> {
+        let tags = repo.get_tags_version_sorted(true);
+
+        for package in packages {
+            // Latest release tag is the first one we find in the descending list of tags
+            let Some(release_tag) =
+                filter_release_tags(tags.iter().map(AsRef::as_ref), &package.name, project).next()
+            else {
+                continue;
+            };
+
+            let temp_dir = self.temp_dir()?;
+
+            let package_store_dir = temp_dir.path().join(&package.name);
+            let package_store_dir = cargo_utils::to_utf8_pathbuf(package_store_dir)
+                .context("temporary directory path is not UTF-8")?;
+
+            // "Download" each package into the temp dir.
+            // We do this by simply creating a new worktree pointing to the release tag.
+            // We could also do this in other ways:
+            // 1. Find relative path to package manifest and checkout package contents
+            //    (see git read-tree and checkout-index) into temp dir
+            // 2. Use `cargo package` to create tarball, and extract it
+            // But the simplest is to use a worktree
+
+            repo.add_worktree(package_store_dir.as_std_path(), release_tag)?;
+
+            let metadata = cargo_utils::get_manifest_metadata(
```

Comment:
Could you please check that the `get_manifest_metadata` `Cargo.toml` path is correct in case of cargo workspace and crates in some nested directory.

Locally, I see that the worktree contains the whole workspace checked out in the `<tmpdir>/<package_name>` and the function is unable to correctly find the package `Cargo.toml` that would be in  `<tmpdir>/<package_name>/<path_to_the_package_in_workspace>/<package_name>/Cargo.toml` .

Reply:
This is a proper path: https://github.com/release-plz/release-plz/pull/1872/files#r1882718672

Reply:
This is intentional - this line loads the manifest of the workspace in the repository root, and [later](https://github.com/release-plz/release-plz/pull/1872/files#diff-4890bc1059a5dffa7f018d59acead0a2e0d874e222fdb10f720f5b3ae1a6d9acR173-R176) finds the right package in the manifest.

Reply:
Good, thanks. I just checked it doesn't work locally in my case when I have several workspaces in different directories.

Reply:
even when using `release-plz --manifest-path`? If so, we should fix it

Reply:
Sorry for that. I've just executed `release-plz` from the subdirectory without `--manifest-path`. If I return back to the repository root and use `--manifest-path`, then everything works. Sorry for the confusion!

Reply:
still, we should have a better error message for it! :)

Reply:
You do bring up a good point though - the code does not support repositories that do not have a `Cargo.toml` in the repository root. I can amend the code to use the same relative path to find the manifest as in the current commit, but like the git tag template, this has the same issue of requiring the relative path of the manifest in the repository to be the same forever. In other words, you can never move the workspace root.

For now, I can make this change and log an error if the manifest root cannot be found in the same relative path; in a follow-up issue, it might be worth exploring the possibility of adding a _repository_ release-plz config (or just a small file) that stores the relative path to the manifest. There are likely better ways of keeping track of where the workspace root is (maybe just find the uppermost `Cargo.toml` file?) but I'll leave it for a later discussion.